lib/spack/spack/variant.py

```diff
@@ -254,116 +239,143 @@ def _flatten(values) -> Collection:
+    def yaml_entry(self) -> Tuple[str, SerializedValueType]:
+        """Returns a (key, value) tuple suitable to be an entry in a yaml dict.
 
         Returns:
-            tuple: values stored in the variant
+            tuple: (name, value_representation)
         """
-        return self._value
-
-    @value.setter
-    def value(self, value: ValueType) -> None:
-        self._value_setter(value)
-
-    def _value_setter(self, value: ValueType) -> None:
-        # Store the original value
-        self._original_value = value
+        if self.type == VariantType.MULTI:
+            return self.name, list(self.values)
+        return self.name, self.values[0]
 
-        if value == "*":
-            self._value = ()
-            return
+    @property
+    def values(self) -> ValueType:
+        return self._values
 
-        if not isinstance(value, (tuple, list)):
-            # Store a tuple of CSV string representations
-            # Tuple is necessary here instead of list because the
-            # values need to be hashed
-            value = tuple(re.split(r"\s*,\s*", str(value)))
+    @property
+    def value(self) -> Union[ValueType, bool, str]:
+        return self._values[0] if self.type != VariantType.MULTI else self._values
+
+    def set(self, *value: Union[bool, str]) -> None:
+        """Set the value(s) of the variant."""
+        if len(value) > 1:
+            value = tuple(sorted(set(value)))
+
+        if self.type != VariantType.MULTI:
+            if len(value) != 1:
+                raise MultipleValuesInExclusiveVariantError(self)
+            unwrapped = value[0]
+            if self.type == VariantType.BOOL and unwrapped not in (True, False):
+                raise ValueError(
+                    f"cannot set a boolean variant to a value that is not a boolean: {unwrapped}"
+                )
 
-        for val in special_variant_values:
-            if val in value and len(value) > 1:
-                msg = "'%s' cannot be combined with other variant" % val
-                msg += " values."
-                raise InvalidVariantValueCombinationError(msg)
+        if "*" in value:
+            raise InvalidVariantValueError("cannot use reserved value '*'")
 
-        # With multi-value variants it is necessary
-        # to remove duplicates and give an order
-        # to a set
-        self._value = tuple(sorted(set(value)))
+        self._values = value
 
     def _cmp_iter(self) -> Iterable:
         yield self.name
         yield self.propagate
-        yield from (str(v) for v in self.value_as_tuple)
+        yield self.concrete
+        yield from (str(v) for v in self.values)
 
-    def copy(self) -> "VariantBase":
-        variant = type(self)(self.name, self._original_value, self.propagate)
-        variant.concrete = self.concrete
-        return variant
+    def copy(self) -> "VariantValue":
+        return type(self)(self.name, self.values, self.type, self.propagate, self.concrete)
```

Comment:
Now that we have a single `VariantValue`, we can use that instead of the indirect `type(self)`